### PR TITLE
[REVIEW] Allow lists_column_wrappers to be constructed from incomplete hierarchies of empty lists.

### PR DESCRIPTION
**Pull request by nvdbaranec**
There was a legitimate list structure that was impossible to declare previously.  Essentially:

`lists_column_wrapper<int> col { {{}}, {} };`

This is a `List<List<int>>` column with 2 rows.  The second row is an empty list, which is legitimate.  However, due to the way the lists_column_wrapper was working internally, when the code was looking to merge the two rows,  `{{}}` and `{}`, it saw the first row as being of type `List<List<int>>` and the second as being of type `List<int>` which was considered a mismatch of the child types (`List` vs. `int`)

The fix is simply to disregard the type of empty columns for the purposes of this merge step.   This also involved a small change in `list::concatenate()` which is used by the lists_column_wrapper internally.

Two other small changes come along with this:

- I retrofitted pre-existing lists tests to use the declaration  `EL` to denote an empty list instead of `LCW{}`.  It reads more clearly.

`lists_column_wrapper<int> col { {EL, {1, 2, 3}}, EL };`

vs

`lists_column_wrapper<int> col { {LCW{}, {1, 2, 3}}, LCW{} };`

I figured now is the time to establish this standard before people more people start using this stuff.

- I added support in column_utilities to display the bitmask of leaf columns when displaying a list.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/tests/copying/concatenate_tests.cu [hunk 1]
```diff
@@ -326,6 +326,9 @@ TEST_F(ListsColumnTest, ConcatenateEmptyLists)
 {
   // to disambiguiate between {} == 0 and {} == List{0}
   using LCW = cudf::test::lists_column_wrapper<int>;
+// clang-format off
+  #define EL LCW{}
+  // clang-format on
```
Comment (davidwendt) [inline cpp/tests/copying/concatenate_tests.cu:331]: What is this for?

File: cpp/tests/utilities_tests/lists_column_wrapper_tests.cpp [hunk 2]
```diff
@@ -904,4 +1059,25 @@ TEST_F(ListColumnWrapperTest, ListOfListOfBools)
...
+  #define EL LCW{}
+  // clang-format on
+
+  // trying to build a column out of a List<List<int>> column, and a List<int> column
+  // is not valid if the leaf lists are not empty.
+  {
+    auto expect_failure = []() { test::lists_column_wrapper<T> list{{{1, 2, 3}}, {4, 5}}; };
+    EXPECT_THROW(expect_failure(), cudf::logic_error);
+  }
 }
```
Comment (davidwendt) [inline cpp/tests/utilities_tests/lists_column_wrapper_tests.cpp:1083]: Empty line needed again :-)

File: cpp/tests/copying/concatenate_tests.cu (hunk 1 above)
Comment (nvdbaranec) [reply cpp/tests/copying/concatenate_tests.cu:331]: The "EL" reads cleaner for "empty list" than "LCW{}".  I retrofitted everything in the various tests to try and set this standard now.

As for clang formatting - it was mangling those #defines in some cases (but only some!), adding a \ and a newline which as causing compile errors.  Not really sure why.

Comment (jrhemstad) [reply cpp/tests/copying/concatenate_tests.cu:331]: Don't use a macro like this. It makes it pretty inscrutable. If you want a special type for an empty list, just make a new type that represents an empty list column wrapper.

**Pull request merged**